src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceManager)), Shared]
+internal class DiagnosticSourceManager : IDiagnosticSourceManager
+{
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _documentProviders;
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _workspaceProviders;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<string> GetSourceNames(bool isDocument)
+        => (isDocument ? _documentProviders : _workspaceProviders).Keys;
+
+    /// <inheritdoc />
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string? sourceName, bool isDocument, CancellationToken cancellationToken)
+    {
+        var providersDictionary = isDocument ? _documentProviders : _workspaceProviders;
+        if (sourceName != null)
+        {
+            Contract.ThrowIfFalse(providersDictionary.TryGetValue(sourceName, out var provider), $"Unrecognized source {sourceName}");
+            return await provider.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            // VS Code (and legacy VS ?) pass null sourceName when requesting all sources.
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var sourcesBuilder);
+            foreach (var kvp in providersDictionary)
+            {
+                // Exclude task diagnostics from the aggregated sources.
+                if (kvp.Key != PullDiagnosticCategories.Task)
```

Comment:
@dibarbet should we also exclude PullDiagnosticCategories.EditAndContinue?

Reply:
I don't think we should specifically exclude either from the aggregated source?  I guess there might be an issue if we dont in VSCode (and start reporting task diags when we don't have a source).  

But maybe we should solve it another way and in the task diagnostic source provider, only return sources when vs diagnostics is enabled.

Reply:
this can probably then be simplified using one of the SelectAsArrayAsync extensions we have
https://github.com/dotnet/roslyn/blob/8a6ff11ba01e3e61eaf117cf120566f8609b6c28/src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs#L447

Reply:
think this applies still.

Reply:
I would need something like SelectManyAsArrayAsync because I need to use `builder.AddRange` rather than `builder.Add`. But cannot figure out how to use SelectManyAsArrayAsync